libchisel/src/verifyimports.rs

```diff
@@ -0,0 +1,898 @@
+                                ValueType::I32,
+                                ValueType::I32,
+                                ValueType::I32,
+                                ValueType::I32,
+                            ],
+                            None,
+                        ),
+                    ),
+                    ImportType::Function(
+                        "ethereum",
+                        "getReturnDataSize",
+                        FunctionType::new(vec![], Some(ValueType::I32)),
+                    ),
+                    ImportType::Function(
+                        "ethereum",
+                        "returnDataCopy",
+                        FunctionType::new(
+                            vec![ValueType::I32, ValueType::I32, ValueType::I32],
+                            None,
+                        ),
+                    ),
+                    ImportType::Function(
+                        "ethereum",
+                        "finish",
+                        FunctionType::new(vec![ValueType::I32, ValueType::I32], None),
+                    ),
+                    ImportType::Function(
+                        "ethereum",
+                        "revert",
+                        FunctionType::new(vec![ValueType::I32, ValueType::I32], None),
+                    ),
+                    ImportType::Function(
+                        "ethereum",
+                        "selfDestruct",
+                        FunctionType::new(vec![ValueType::I32], None),
+                    ),
+                ].iter()
+                .cloned()
+                .collect(),
+                require_all: false,
+                allow_unlisted: false,
+            }),
+            _ => Err(()),
+        }
+    }
+
+    // Utility functions used in tests to get more coverage
+    #[cfg(test)]
+    fn set_require_all(&mut self, arg: bool) {
+        self.require_all = arg;
+    }
+
+    #[cfg(test)]
+    fn set_allow_unlisted(&mut self, arg: bool) {
+        self.allow_unlisted = arg;
+    }
+}
+
+impl<'a> ModuleValidator for VerifyImports<'a> {
+    fn validate(self, module: &Module) -> Result<bool, String> {
+        let import_section_len = if let Some(section) = module.import_section() {
+            section.entries().len()
+        } else {
+            0
+        };
+
+        Ok(match (self.require_all, self.allow_unlisted) {
+            (true, true) => self
+                .entries
+                .iter()
+                .map(|e| e.is_imported(module))
+                .find(|e| *e == false)
+                .is_none(),
+            (true, false) => {
+                self
+                    .entries
+                    .iter()
+                    .map(|e| e.is_imported(module))
+                    .find(|e| *e == false)
+                    .is_none()
+                    && (self.entries.len() == import_section_len)
+            }
+            (false, true) => self
+                .entries
```

Comment:
This case basically doesn't care what the imports are, except that the listed ones must match.